In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv("./data/heart_2022_with_nans.csv")
df.head(10)

NameError: name 'pd' is not defined

In [ ]:
df.info() #Hay variables tanto categóricas, como numéricas como booleanas.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445132 entries, 0 to 445131
Data columns (total 40 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   State                      445132 non-null  object 
 1   Sex                        445132 non-null  object 
 2   GeneralHealth              443934 non-null  object 
 3   PhysicalHealthDays         434205 non-null  float64
 4   MentalHealthDays           436065 non-null  float64
 5   LastCheckupTime            436824 non-null  object 
 6   PhysicalActivities         444039 non-null  object 
 7   SleepHours                 439679 non-null  float64
 8   RemovedTeeth               433772 non-null  object 
 9   HadHeartAttack             442067 non-null  object 
 10  HadAngina                  440727 non-null  object 
 11  HadStroke                  443575 non-null  object 
 12  HadAsthma                  443359 non-null  object 
 13  HadSkinCancer              44

In [ ]:
df.shape

(445132, 40)

In [ ]:
df.columns

Index(['State', 'Sex', 'GeneralHealth', 'PhysicalHealthDays',
       'MentalHealthDays', 'LastCheckupTime', 'PhysicalActivities',
       'SleepHours', 'RemovedTeeth', 'HadHeartAttack', 'HadAngina',
       'HadStroke', 'HadAsthma', 'HadSkinCancer', 'HadCOPD',
       'HadDepressiveDisorder', 'HadKidneyDisease', 'HadArthritis',
       'HadDiabetes', 'DeafOrHardOfHearing', 'BlindOrVisionDifficulty',
       'DifficultyConcentrating', 'DifficultyWalking',
       'DifficultyDressingBathing', 'DifficultyErrands', 'SmokerStatus',
       'ECigaretteUsage', 'ChestScan', 'RaceEthnicityCategory', 'AgeCategory',
       'HeightInMeters', 'WeightInKilograms', 'BMI', 'AlcoholDrinkers',
       'HIVTesting', 'FluVaxLast12', 'PneumoVaxEver', 'TetanusLast10Tdap',
       'HighRiskLastYear', 'CovidPos'],
      dtype='object')

In [6]:
df["AgeCategory"].value_counts()

AgeCategory
Age 65 to 69       47099
Age 60 to 64       44511
Age 70 to 74       43472
Age 55 to 59       36821
Age 80 or older    36251
Age 50 to 54       33644
Age 75 to 79       32518
Age 40 to 44       29942
Age 45 to 49       28531
Age 35 to 39       28526
Age 18 to 24       26941
Age 30 to 34       25807
Age 25 to 29       21990
Name: count, dtype: int64

In [ ]:
print (df["Sex"].value_counts())
print (df["Sex"].value_counts(normalize=True)*100)

Sex
Female    235893
Male      209239
Name: count, dtype: int64
Sex
Female    52.993943
Male      47.006057
Name: proportion, dtype: float64


In [ ]:
df["RaceEthnicityCategory"].value_counts()

RaceEthnicityCategory
White only, Non-Hispanic         320421
Hispanic                          42917
Black only, Non-Hispanic          35446
Other race only, Non-Hispanic     22713
Multiracial, Non-Hispanic          9578
Name: count, dtype: int64

|Columna|Descripción|Tipo|Categorías|
|:-----:|:-----:|:-----:|:-----:|
|State|Estado de residencia de la persona encuestada en USA|Categórica|Estados de USA|
|Sex|Género de la persona encuestada|Categórica|Male,Female|
|General Health|Estado general de la salud|Categórica|Poor,Fair,Good,Very good,Excellent|
|Physical Health Days|Días con mala salud física de los últimos 30 días|Numérica|Días(0-30)|
|Mental Health Days|Días con mala salud mental de los últimos 30 días|Numérica|Días(0-30)|
|Last Checkup Time|Tiempo desde el último chequeo|Categórica|<1 year, 1-2 years, >5 years, Never|
|Physical Activities|Si realiza actividad física o no|Booleana|Yes/No|
|Sleep Hours|Horas de sueño diario|Numérica|Horas|
|Removed Teeth|Cantidad de dientes quitados|Categórica|NaN,1-5,6+,All|
|Had Heart Attach|Si ha sufrido o no un ataque al corazón|Booleana|Yes/No|
|Had Angina|Si ha sufrido una angina de pecho|Booleana|Yes/No|
|Had Stroke|Si ha sufrido un derrame cerebral|Booleana|Yes/No|
|Had Asthma|Si tiene asma o no|Booleana|Yes/No|
|Had Skin Cancer|Si ha sufrido cáncer de piel|Booleana|Yes/No|
|Had COPD|Si tiene EPOC|Booleana|Yes/No|
|Had Depressiva Disorder|Si ha sido diagnosticado de trastorno depresivo|Booleana|Yes/No|
|Had Kidney Disease|Si tiene problemas renales|Booleana|Yes/No|
|Had Arthritis|Diagnóstico de artritis|Booleana|Yes/No|
|Had Diabetes|Diagnóstico de diabetes|Categórica|Yes, No, Borderline, During Pregnancy|
|Deaf Or Hard Of Hearing|Problemas auditivos|Booleana|Yes/No|
|Blind Or Vision Difficulty|Problemas visuales|Booleana|Yes/No|
|Difficulty Concentrating|Dificultad para concentrarse|Booleana|Yes/No|
|Difficulty Walking|Dificultad para andar|Booleana|Yes/No|
|Difficulty Dressing Bathing|Dificultad para vestirse o bañarse|Booleana|Yes/No|
|Difficulty Errands|Dificultad para realizar recados solo|Booleana|Yes/No|
|Smoker Status|Estado respecto al hábito de fumar|Categórica|Never, Former, Current|
|ECigarette Usage|Uso de cigarrillos electrónicos|Categórica|Never, Former, Current|
|Chest Scan|Se se ha realizado un escáner torácico|Booleana|Yes/No|
|Race Ethnicity Category|Grupo racial o étnico|Categórica|White, Black, Hispanic, etc.|
|Age Category|Edad de la persona encuestada|Numérica|Rangos de edad|
|Height In Meters|Altura de la persona encuestada|Numérica|Metros|
|Weight In Kilograms|Peso de la persona encuestada|Numérica|kg|
|BMI|Índice de Masa Corporal|Numérica|kg|
|Alcohol Drinkers|Consumo de alcohol|Booleana|Yes/No|
|HIV Testing|Si se ha realizado el test de VIH|Booleana|Yes/No|
|Flu Vax Last 12|Si se ha puesto la vacuna de la gripe en los últimos 12 meses|Booleana|Yes/No|
|Pneumo Vax Ever|Si le han puesto la vacuna neumocócica alguna vez|Booleana|Yes/No|
|Tetanus Last 10 Tdap|Si le han puesto la vacuna antitetánica (Tdap) en los últimos 10 años|Categórica|Yes, No, Don't know|
|High Risk Last Year|Conductas de alto riesgo durante el último año|Booleana|Yes/No|
|CovidPos|Resultado positivo en una prueba de COVID-19|Categórica|Yes, No, Don't know|



In [ ]:
total_nulos = df.isna().sum().sum()
total_nulos

np.int64(902665)

In [ ]:
print ((df.isna().sum().sort_values(ascending=False)))


TetanusLast10Tdap            82516
PneumoVaxEver                77040
HIVTesting                   66127
ChestScan                    56046
CovidPos                     50764
HighRiskLastYear             50623
BMI                          48806
FluVaxLast12                 47121
AlcoholDrinkers              46574
WeightInKilograms            42078
ECigaretteUsage              35660
SmokerStatus                 35462
HeightInMeters               28652
DifficultyErrands            25656
DifficultyConcentrating      24240
DifficultyWalking            24012
DifficultyDressingBathing    23915
BlindOrVisionDifficulty      21564
DeafOrHardOfHearing          20647
RaceEthnicityCategory        14057
RemovedTeeth                 11360
PhysicalHealthDays           10927
AgeCategory                   9079
MentalHealthDays              9067
LastCheckupTime               8308
SleepHours                    5453
HadAngina                     4405
HadSkinCancer                 3143
HadHeartAttack      

In [ ]:
total_celdas = df.shape[0]*df.shape[1]
porcentaje_nulos = (total_nulos/total_celdas)*100
round (porcentaje_nulos,2)

np.float64(5.07)

In [ ]:
print (df.duplicated().sum())

157


# **Notas**

- En este primer análisis del dataset, podemos observar que hay un total de 40 columnas y 445.132 filas.

- El total de valores nulos es de 889.788. La categoría que tiene más nulos es la de Tetanus Last 10Tdap.

- De estos valores nulos (NaN), tenemos que decidir qué hacer con ellos, ya que sería interesante reemplazar solo algunos valores por "UNK". De otros, como por ejemplo los que están en las variables numéricas, podríamos calcularlo en base a la mediana.

- Encontramos un total de 157 valores duplicados, por lo que vamos a decir qué vamos a hacer con ellos.

- El porcentaje de respuesta entre hombres y mujeres está muy igualado (53% vs 47%)

## Detectar errores y cambiarlos

- Muchas de las columnas donde hay valores numéricos (Sleep Hours, Age Category, Physical Health Days, Mental Health Days, etc. son booleanos. Quitar el 0 y dejarlo como un entero)

- Los nombres de las columnas no tienen espacios. Sería ideal añadirle ese espacio.